### Four main types of numbers:

1. Integer numbers (Z)  - Python Type: int
2. Rational numbers (Q) - Python Type: fractions.Fraction
3. Real numbers (R)     - Python Type: float, decimal.Decimal
4. Complex numbers (C)  - Python Type: complex

Remember: Z is subset of Q is subset of R is subset of C --> C is the biggest subset

### In Python, there are Five main types of numbers:

1. Boolean truth values - Python Type: bool
2. Integer numbers (Z)  - Python Type: int
3. Rational numbers (Q) - Python Type: fractions.Fraction
4. Real numbers (R)     - Python Type: float, decimal.Decimal
5. Complex numbers (C)  - Python Type: complex

In [1]:
# what the large magitude
a = int('10011', 2) # base 2
b = int('19', 10) # base 10

print(a, b)

19 19


### a and b are presented above 
how large can a Python int become (posivive or negative)?

decimal number 19 required 5 bits (10011) in order to store or represent it

In [2]:
# what is the largeest integer number can be presented using 8 bits?

c = int('11111111', 2) # base 2
c


255

### the largest number that we can use is 255
255 = 2**8 - 1

In [4]:
2**8 - 1

255

In [8]:
# now let check for the negative number

d = int('-1111111', 2) # base 2 ( 1 bit is reserved to represent the sign of number)
d


-127

### The largest number that we can represent using 7 bits is 127
2**7 - 1 = 127

In [9]:
2**7 - 1

127

## So, using 8 bits we can represent all the integers in the range of [-127, 127]

and 0 dese not require a sign then we can end up with the range of [-128, 127]

### [-2^7, 2^7 - 1]

### using 16 bits to store (signed) integers
2^(16-1) = 2^15 Range is [-32768 ... 32767]

### using 32 bits to store (signed) integers
2^(32-1) = 2^31 Range is [-2^31 ... 2^31 -1 ]


In [10]:
-2**31

-2147483648

In [14]:
2**31 - 1

2147483647

### Memory spaces (bytes) are limited by the address number
such as in a 32-bit OS, if we had an unsigned integer type our range is [0, 2**32] 

In [15]:
# how many bytes od addressable memory of 32-bit OS
print(2**32, 'bytes')
print(2**32/1024, 'kB')
print(2**32/1024/1024, 'MB')
print(2**32/1024/1024/1024, 'GB')

4294967296 bytes
4194304.0 kB
4096.0 MB
4.0 GB


### so for 32-bit OS we require to add at least 4GB RAM but we are not going to be able to address all of that memory

In [16]:
print(type(100))

<class 'int'>


In [17]:
import sys

sys.getsizeof(0)

24

In [18]:
# it takes up 24 bytes for 0

In [19]:
sys.getsizeof(1)

28

In [20]:
# let's check the big number

sys.getsizeof(2**1000)

160

In [21]:
# how many bits did we use for the actual number
# 160: actual number
# 24: number of bytes
# 8: number of bits

(160 - 24) * 8

1088

In [22]:
# so we use 1088 bits to store the large number of 2**1000 

2**1000

10715086071862673209484250490600018105614048117055336074437503883703510511249361224931983788156958581275946729175531468251871452856923140435984577574698574803934567774824230985421074605062371141877954182153046474983581941267398767559165543946077062914571196477686542167660429831652624386837205668069376